lib/random.py

```diff
@@ -0,0 +1,187 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Generate pseudo random numbers. Should not be used for security purposes."""
+
+from __go__.math.rand import Uint32, Seed
+from __go__.math import Pow as _pow
+from __go__.time import Now
+
+
+BPF = 53  # Number of bits in a float
+RECIP_BPF = _pow(2, -BPF)
+
+
+def _gorandom(nbytes):
+  byte_arr = []
+  while len(byte_arr) < nbytes:
+    i = Uint32()
+    byte_arr.append(i & 0xff)
+    byte_arr.append(i >> 8 & 0xff)
+    byte_arr.append(i >> 16 & 0xff)
+    byte_arr.append(i >> 24 & 0xff)
+  byte_arr = byte_arr[0:nbytes]
+  return byte_arr
+
+
+def _notimplemented(*args, **kwargs):
+  raise NotImplementedError
+
+
+# This is a slow replacement for int.bit_length.
+# We should stop using this if it is implemented.
+def _int_bit_length(n):
+  bits = 0
+  while n:  # 1 bit steps
+    n = n / 2
+    bits += 1
+  return bits
+
+
+# Replacement for int.from_bytes (big endian)
+def _int_from_bytes(bytes):
+  i = 0
+  n = len(bytes) - 1
+  while n >= 0:
+    i += bytes[n] << (8 * n)
+    n -= 1
+  return i
+
+
+class GrumpyRandom(object):
+  """Alternate random number generator using math.rand.read as a replacement
+  for urandom. Implemented like SystemRandom from cPythons random.py lib.
+  """
+
+  def random(self):
+    """Get the next random number in the range [0.0, 1.0)."""
+    return (_int_from_bytes(_gorandom(7)) >> 3) * RECIP_BPF
+
+  def getrandbits(self, k):
+    """getrandbits(k) -> x.  Generates an int with k random bits."""
+    if k <= 0:
+      raise ValueError('number of bits must be greater than zero')
+    if k != int(k):
+      raise TypeError('number of bits should be an integer')
+    numbytes = (k + 7) // 8                       # bits / 8 and rounded up
+    x = _int_from_bytes(_gorandom(numbytes))
+    return x >> (numbytes * 8 - k)                # trim excess bits
+
+  def seed(self, a=None):
+    """Seed the golang.math.rand generator."""
+    if a is None:
+      a = Now().UnixNano()
+    Seed(a)
+
+  def _randbelow(self, n):
+    """Return a random int in the range [0,n)."""
+    # change once int.bit_length is implemented.
+    # k = n.bit_length()
+    k = _int_bit_length(n)
+    r = getrandbits(k)
+    while r >= n:
+      r = getrandbits(k)
+    return r
+
+  def getstate(self):
+    raise NotImplementedError('Entropy source does not have state.')
+
+  def setstate(self):
+    raise NotImplementedError('Entropy source does not have state.')
+
+
+# Most of the following code is taken from the cpython std lib.
```

Comment:
This is unfortunately a licensing problem because PSF licensed code needs to reside in third_party/stdlib. Is there a way we can reuse the CPython code as-is?